Add model manager for machine learning models

idlir-shkurti-frequenz opened the pull request:
This PR adds a simple model manager module which implements the logic needed to load, update and monitor machine learning models.
Currently there are two model repository structures: 
- SingleModelRepository - This is a structure where we direct our handler to a repository with a single model in it.
- DirectModelRepository - This repository structure takes in dictionary input with a key which is the name of the model and a value which is the  direct path to each specific module.

This PR is heavily based (but slimmed down) on https://github.com/frequenz-floss/frequenz-sdk-python/pull/397

idlir-shkurti-frequenz committed:
Add model manager as BackgroundService for handling machine learning models

Signed-off-by: Idlir Shkurti <idlir.shkurti@frequenz.com>

llucax commented:
Maybe this was talked elsewhere, but it might be nice to provide some extra context about what changed between the closing of #397 and now. Was the experiment inside the actor successful and we think this new slimmed down implementation is good/useful enough to go in the SDK?

idlir-shkurti-frequenz commented:
> Maybe this was talked elsewhere, but it might be nice to provide some extra context about what changed between the closing of #397 and now. Was the experiment inside the actor successful and we think this new slimmed down implementation is good/useful enough to go in the SDK?

@llucax We decided that it would be best to start the model loading immediately inside the sdk so that we don't have to migrate extra code later. This slimmed down version should be good enough for the initial usecases. We don't need extra `weeday` and `enumerated` repositories but only a way to load and monitor ML models from a pre-defined path.

llucax commented on src/frequenz/sdk/model/__init__.py:
```diff
@@ -0,0 +1,19 @@
+# License: MIT
+# Copyright © 2023 Frequenz Energy-as-a-Service GmbH
```
Nitpick: I guess this is 2023 because the code it was taken from was from 2023, right?

idlir-shkurti-frequenz replied:
You are right. I will change it

llucax replied:
Either way is fine with me.

llucax commented on src/frequenz/sdk/model/_manager.py:
```diff
@@ -0,0 +1,273 @@
+
+
+@dataclass
+class SingleModelRepository(ModelRepository):
```
```suggestion
class SingleModelRepository(ModelRepository[str]):
```

llucax commented on src/frequenz/sdk/model/_manager.py:
```diff
@@ -0,0 +1,273 @@
+class SingleModelRepository(ModelRepository):
+    """Repository to manage a single model."""
+
+    model_path: str
```
```suggestion
    model_path: pathlib.Path
```

llucax commented on src/frequenz/sdk/model/_manager.py:
```diff
@@ -0,0 +1,273 @@
+    model_path: str
+    """The path to a single model file."""
+
+    def _get_allowed_keys(self) -> list[Any]:
```
```suggestion
    def _get_allowed_keys(self) -> collections.abc.Set[str]:
```

llucax commented on src/frequenz/sdk/model/_manager.py:
```diff
@@ -0,0 +1,273 @@
+"""The model type."""
+
+
+class ModelRepository(abc.ABC):
```
Can't we parametrize keys in this class with `T` too? I think we should avoid using `Any` as it completely breaks the type system.

```suggestion
class ModelRepository(abc.ABC, Generic[T]):
```

llucax commented on src/frequenz/sdk/model/_manager.py:
```diff
@@ -0,0 +1,273 @@
+    """
+
+    @abc.abstractmethod
+    def _get_allowed_keys(self) -> list[Any]:
```
```suggestion
    def _get_allowed_keys(self) -> collections.abc.Set[T]:
```

llucax commented on src/frequenz/sdk/model/_manager.py:
```diff
@@ -0,0 +1,273 @@
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def _build_path(self, key: Any) -> str:
```
I would use `pathlib.Path` instead of `str` for paths.

```suggestion
    def _build_path(self, key: T) -> pathlib.Path:
```

llucax commented on src/frequenz/sdk/model/_manager.py:
```diff
@@ -0,0 +1,273 @@
+        """
+        raise NotImplementedError
+
+    def get_models_config(self) -> dict[Any, str]:
```
```suggestion
    def get_models_config(self) -> dict[T, pathlib.Path]:
```

llucax commented on src/frequenz/sdk/model/_manager.py:
```diff
@@ -0,0 +1,273 @@
+        """
+        return ["Single"]
+
+    def _build_path(self, key: Any) -> str:
```
```suggestion
    def _build_path(self, key: str) -> pathlib.Path:
```

llucax commented on src/frequenz/sdk/model/_manager.py:
```diff
@@ -0,0 +1,273 @@
+
+
+@dataclass
+class DirectModelRepository(ModelRepository):
```
```suggestion
class DirectModelRepository(ModelRepository[str]):
```

Also, I'm not sure about the name, why direct? This comes from when we had repos for weekday, etc., right?

llucax commented on src/frequenz/sdk/model/_manager.py:
```diff
@@ -0,0 +1,273 @@
+class DirectModelRepository(ModelRepository):
+    """Repository to manage models based on a direct configuration dictionary."""
+
+    model_config: dict[str, str]
```
```suggestion
    model_config: dict[str, pathlib.Path]
```

llucax commented on src/frequenz/sdk/model/_manager.py:
```diff
@@ -0,0 +1,273 @@
+    model_config: dict[str, str]
+    """A dictionary mapping keys to the full paths of model files."""
+
+    def _get_allowed_keys(self) -> list[str]:
```
```suggestion
    def _get_allowed_keys(self) -> collections.abc.Set[str]:
```

llucax commented on src/frequenz/sdk/model/_manager.py:
```diff
@@ -0,0 +1,273 @@
+            A list of keys as specified in the model_config dictionary.
+        """
+        assert self.model_config, "Model configuration must be set."
+        return list(self.model_config.keys())
```
```suggestion
        return frozenset(self.model_config.keys())
```

llucax commented on src/frequenz/sdk/model/_manager.py:
```diff
@@ -0,0 +1,273 @@
+        assert self.model_config, "Model configuration must be set."
+        return list(self.model_config.keys())
+
+    def _build_path(self, key: str) -> str:
```
```suggestion
    def _build_path(self, key: str) -> pathlib.Path:
```

llucax commented on src/frequenz/sdk/model/_manager.py:
```diff
@@ -0,0 +1,273 @@
+            model_repository: a model repository that defines the mapping
+                between model keys and paths.
+        """
+        self._models: dict[Any, _Model[T]] = {}
```
Here we could add another parameter, like `U`, for the key type, or use `object` if we want to make it an opaque type. The difference between using `object` and `Any` is `Any` will pass any type checking, like `key.i_dont_exist`, while object will raise an error. Since the model repository has a typed key, I think it makes more sense to go with the extra type parameter.
```suggestion
        self._models: dict[U, _Model[T]] = {}
```

llucax commented on src/frequenz/sdk/model/_manager.py:
```diff
@@ -0,0 +1,273 @@
+    """The path to the model file."""
+
+
+class ModelManager(Generic[T]):
```
```suggestion
class ModelManager(Generic[T, U]):
```

llucax commented on src/frequenz/sdk/model/_manager.py:
```diff
@@ -0,0 +1,273 @@
+            self._start_model_monitor()
+        )
+
+    def __getitem__(self, key: Any) -> T:
```
```suggestion
    def __getitem__(self, key: U) -> T:
```

llucax commented on src/frequenz/sdk/model/_manager.py:
```diff
@@ -0,0 +1,273 @@
+        """
+        return self.get_model(key)
+
+    def get_model(self, key: Any = None) -> T:
```
To be correct, this needs to get a bit more complicated if we make keys generic. We should use a sentinel class, like:

```py
class Unspecified:
    pass
```

```suggestion
    def get_model(self, key: U = Unspecified) -> T:
```

llucax commented on src/frequenz/sdk/model/_manager.py:
```diff
@@ -0,0 +1,273 @@
+
+        return self._models[key].data
+
+    def get_paths(self) -> list[Path | str]:
```
I would return an iterator here, as it is probably all we need and makes the interface more flexible. Also I would keep it simple and only return `pathlib.Path`:
```suggestion
    def get_paths(self) -> Iterator[Path]:
```

llucax commented on src/frequenz/sdk/model/_manager.py:
```diff
@@ -0,0 +1,273 @@
+        if key is None:
+            if len(self._models) != 1:
+                raise KeyError("No key provided for multi-model retrieval")
+            key = list(self._models.keys())[0]
```
```suggestion
        if key is Unspecified:
            if len(self._models) != 1:
                raise KeyError("A key must be provided")
            key = list(self._models.keys())[0]
```

llucax commented on src/frequenz/sdk/model/_manager.py:
```diff
@@ -0,0 +1,273 @@
+        Returns:
+            the paths of all loaded models.
+        """
+        return [model.path for model in self._models.values()]
```
This returns an iterator instead of a list:
```suggestion
        return (model.path for model in self._models.values())
```

llucax reviewed (commented):
I'm not crazy about the `get_model()` interface because it is too dynamic, it would be good if it could be more static and misuse would be detected by `mypy` instead of at runtime.

Looking at the bigger picture here, it seems this code is still inheriting a lot of the complexity of the previous PR. Do we really need a special case for a single model repository? Can't we just use the dict-based repo and let user pass a key with the value they like the most? It seems like a lot of added extra complexity just to avoid passing a key.

I would just remove the special case here and also the `SingleModelRepository` class and force users to pass a key. Once this is done, I don't think it makes sense to keep the abstract base class `ModelRepository`, and if we remove it, we can just rename `DirectModelRepository` to `ModelRepository`. This will make this PR much much simpler.

If we need more abstraction and more types of repositories in the future, we can bring the complexity back, but I wouldn't bet on it just yet.

idlir-shkurti-frequenz commented:
> Looking at the bigger picture here, it seems this code is still inheriting a lot of the complexity of the previous PR. Do we really need a special case for a single model repository? Can't we just use the dict-based repo and let user pass a key with the value they like the most? It seems like a lot of added extra complexity just to avoid passing a key.

That is actually something I was also considering. @daniel-zullo-frequenz @cwasicki I think I agree with Luca here that we can just have the one repo structure with a dict input and forget about the single model repo structure. If that is ok with you I will make the change.

cwasicki commented:
>If that is ok with you I will make the change.

Agree. I think we can also limit it to the manager and don't need the repositories.

cwasicki commented:
I suggest for now to only implement the bare minimum functionality that we currently need.

llucax commented:
> Agree. I think we can also limit it to the manager and don't need the repositories.

I think having only the "dict repository" is simple enough and gives more flexibility to separate different models if we need to, also the manager and repository are very closely related, so doing this would basically mean throwing everything and start again, right?

IMHO having `DirectModelRepository` (as `ModelRepository`) and the `ModelManager` that can monitor more than one `ModelRepository` strikes a good balance between simplicity and flexibility. But if you think that is still over-bloated for our current needs, I'm also OK to remove the repositories altogether.

idlir-shkurti-frequenz commented:
@llucax @cwasicki I made some changes to the PR. I removed the repository structures for simplicity and changed the `ModelManager` to require a `model_config` in its initialisation. This `model_config` is simply a `dict[str, Path]` pointing to the models. 

@llucax I changed the key from `Any` to `str` and not as a generic like you proposed. Do you think it is important for now to make them generic? This will simply be the name of the model and I would assume as key is good enough for now. If you think it still makes sense to change them to generic then let me know.

llucax commented:
> I changed the key from Any to str and not as a generic like you proposed. Do you think it is important for now to make them generic?

TL;DR: I think it is OK. It was a journey to convince myself of that, so I will keep the long version for future reference :D

I was going to say it's not important in this context, but then I was thinking about the case where we want to use an enum for safety, as `str` is very prone to errors due to typos for example, which can't be detected by linters or `mypy`, so they will only occur at runtime.

This is why I still liked the separation between the repository and the manager, because then you can have a "weekly" model, where keys are a enum with the week days, and then you can have an arbitrary model with some `str` keys that you build more dynamically.

I think it would be good to encourage users to design their code with safety in mind, so I'm not sure. I understand this is good enough for the current use cases but I feel it might be a move in the wrong direction.

But I'm not sure if it is worth to provide a model repository that it is just a `dict` with `str` keys, because then most users will just use that and will have the same issues. So it probably makes sense to add the safety once we want to provide safer repositories, like a day of the week repo, where keys can only be members of a `DayOfTheWeek` enum for example, etc., as we discussed originally.

llucax commented on src/frequenz/sdk/model/_manager.py:
```diff
@@ -0,0 +1,98 @@
+class ModelManager(Generic[T]):
+    """Manage machine learning models with direct dictionary configuration."""
+
+    def __init__(self, model_config: dict[str, Path]) -> None:
```
I would still call the mapping of model to path a *repository*.

```suggestion
    def __init__(self, repository: dict[str, Path]) -> None:
```

Or maybe `model_paths`, but `config` sounds weird in this context, for me it sounds like it would configure how often they are checked for changes or stuff like that, configuration for the manager itself.

idlir-shkurti-frequenz replied:
I think `model_paths` is better. I will change it to that.

llucax commented on src/frequenz/sdk/model/_manager.py:
```diff
@@ -0,0 +1,98 @@
+                self.reload_model(Path(event.path))
+
+    def reload_model(self, model_path: Path) -> None:
+        """Reload a model from the given path."""
+        for key, model in self._models.items():
+            if model.path == model_path:
+                model.data = self._load(model_path)
+                return
+        raise ValueError(f"No model found with path: {model_path}")
```
I wouldn't raise an exception here, otherwise one error will make the manager stop processing any other change notifications.

Even more, I think there could be races here where you might get a create notification for example, but the user removed it immediately so the file will not exist when you want to open it, so I would add a `try/except` block here catching *at least* any `OSError`, logging the error and returning, so the loop can continue processing events.

idlir-shkurti-frequenz replied:
Done

llucax commented on src/frequenz/sdk/model/_manager.py:
```diff
@@ -0,0 +1,98 @@
+                return
+        raise ValueError(f"No model found with path: {model_path}")
+
+    async def join(self) -> None:
+        """Await the monitoring task, and return when the task completes."""
+        if self._monitoring_task and not self._monitoring_task.done():
+            await self._monitoring_task
+
+    async def _stop_model_monitor(self) -> None:
+        """Stop monitoring the model paths for changes."""
+        if self._monitoring_task:
+            await cancel_and_await(self._monitoring_task)
```
Oh, this makes me think this class should be a [`BackgroundService`](https://frequenz-floss.github.io/frequenz-sdk-python/v0.25/reference/frequenz/sdk/actor/#frequenz.sdk.actor.BackgroundService) instead, then you will get the `join()` and `stop()` methods automatically, and we keep a consistent interface.

idlir-shkurti-frequenz replied:
Just to clarify we wouldn't have `join()` right? We would have `wait()` instead?

idlir-shkurti-frequenz replied:
Just to clarify we wouldn't have `join()` right? We would use `wait()` instead?

idlir-shkurti-frequenz replied:
Ok I just read a bit more on `BackgroundService`. We wouldn't even need to implement it at all I see since it is already taken care of.

llucax replied:
Exactly. You would be able to use it as an async context manager for example (`async with ModelManager(...):`) or await on it directly (`await model_manager`).

idlir-shkurti-frequenz replied:
Done!

llucax commented on tests/model/test_model_manager.py:
```diff
@@ -0,0 +1,111 @@
+
+    # Map file paths to their respective mock file objects with corresponding data
+    mock_files = {
+        "path/to/model1.pkl": mock_open(read_data=pickled_model1),
+        "path/to/model2.pkl": mock_open(read_data=pickled_model2),
```
Interesting, I didn't know this `mock_open` function.

llucax reviewed (commented):
I got a bit lost in the patching and mock files in the tests, so I will have a look at it later.

cwasicki commented on src/frequenz/sdk/model/_manager.py:
```diff
@@ -0,0 +1,98 @@
+    async def _start_model_monitor(self) -> None:
+        """Start monitoring the model paths for changes."""
+        model_paths = [model.path for model in self._models.values()]
+        file_watcher = FileWatcher(paths=[str(path) for path in model_paths])
```
The `str` should not be required.

cwasicki commented on src/frequenz/sdk/model/_manager.py:
```diff
@@ -0,0 +1,98 @@
+            return cast(T, model)
+
+    def get_model(self, key: str) -> T:
+        """Retrieve a loaded model by key."""
```
The doc in this module is missing for arguments and return values.

llucax replied:
This is allowed by the google style guide provided that you use `Return ...` and it is clear enough from the summary what the arguments should be. In this case I think it is equivalent, but the tool might require `Return ...` instead to adhere more strictly with the google style, so I would update it. Still this probably needs to document that it raises an exception if the key is not found, so it might be better to do the full doc here.

cwasicki replied:
Oh, I didn't know that. In this case I would be fine with this, if there is one place that explains in a bit more detail what the model manager is doing. E.g. things like unique keys, model pathes, what happens if a model is reloaded.

idlir-shkurti-frequenz replied:
I extended the docstring here with the arguments and return value

cwasicki commented on src/frequenz/sdk/model/_manager.py:
```diff
@@ -0,0 +1,98 @@
+        for key, model in self._models.items():
+            if model.path == model_path:
+                model.data = self._load(model_path)
+                return
```
This assumes that each path exists only for one key. Not sure if we want to be so strict about this.

cwasicki replied:
If we do, we should verify this somewhere and also document it.

cwasicki commented on src/frequenz/sdk/model/_manager.py:
```diff
@@ -0,0 +1,98 @@
+class _Model(Generic[T]):
+    """Represent a machine learning model."""
+
+    data: T
```
I wonder whether we want to be more specific about the model implementation. For instance the `data` will most likely be more complex, e.g. a dictionary that holds a model, feature names and other metadata.

In the current version, there is nothing in the `_Model` neither in the `ModelManager` that is "model"-specific. In fact this module is a `PickleManager`. Don't think it's a bad thing, but maybe we could be more clear about this if we keep the generic approach.

llucax replied:
I agree but think we don't need to do it right now. Not calling it `PickleManager` allows us to grow in time with model-specific stuff.

cwasicki commented on src/frequenz/sdk/model/_manager.py:
Maybe a more generic namespace like `ml` is better suited? That would also cover other things like the features module that we want to move to the SDK.

cwasicki commented on src/frequenz/sdk/ml/_model_manager.py:
```diff
@@ -0,0 +1,136 @@
+        for key, path in self.model_paths.items():
+            self._models[key] = _Model(data=self._load(path), path=path)
+
+    def _load(self, path: Path) -> T:
```
This can be made a static function

idlir-shkurti-frequenz replied:
Done!

llucax commented on src/frequenz/sdk/ml/_model_manager.py:
```diff
@@ -0,0 +1,137 @@
+class _Model(Generic[T]):
+    """Represent a machine learning model."""
+
+    data: T
+    path: Path
```
Ideally these members should be documented, but I guess being internal and so simple it is OK to skip.

llucax commented on src/frequenz/sdk/ml/_model_manager.py:
```diff
@@ -0,0 +1,137 @@
+        Args:
+            model_paths: A dictionary of model keys and their corresponding file paths.
+        """
+        super().__init__()
```
I recommend passing a `name` here (and take an optional name in the constructor too). It is used to print some logs by `BackgroundService`.

idlir-shkurti-frequenz replied:
Done!

llucax replied:
You still need to forward the name to the `BackgroundService`.
```suggestion
        super().__init__(name=name)
```

llucax commented on src/frequenz/sdk/ml/_model_manager.py:
```diff
@@ -0,0 +1,137 @@
+        """
+        if not path.exists():
+            raise FileNotFoundError(f"The model path {path} does not exist.")
+        with path.open("rb") as file:
+            return cast(T, pickle.load(file))
```
Oh damn, I just realized this file reading is not async and can potentially block the event loop for a non trivial amount of time, which is not ideal. Python's built-in `asyncio` doesn't support async file I/O, so we would need to use another library for this (like [aiofiles](https://pypi.org/project/aiofiles/)). I created a separate issue because this is way beyond this PR:

* https://github.com/frequenz-floss/frequenz-sdk-python/issues/929

llucax commented on src/frequenz/sdk/ml/_model_manager.py:
```diff
@@ -0,0 +1,137 @@
+        with path.open("rb") as file:
+            return cast(T, pickle.load(file))
+
+    def start(self) -> None:
```
```suggestion
    @override
    def start(self) -> None:
```

llucax commented on src/frequenz/sdk/ml/_model_manager.py:
```diff
@@ -0,0 +1,137 @@
+            self._tasks.add(task)
+            _logger.info("Started ModelManager service with task %s", task)
+
+    async def run(self) -> None:
```
```suggestion
    @override
    async def run(self) -> None:
```

llucax commented on src/frequenz/sdk/ml/_model_manager.py:
```diff
@@ -0,0 +1,137 @@
+    async def run(self) -> None:
+        """Monitor model file paths and reload models as necessary."""
+        model_paths = [model.path for model in self._models.values()]
+        file_watcher = FileWatcher(paths=list(model_paths))
```
```suggestion
        file_watcher = FileWatcher(paths=model_paths)
```

This is already a list (when using `[...]` a list is produced, when using `(...)` a generator is produced).

llucax commented on src/frequenz/sdk/ml/_model_manager.py:
```diff
@@ -0,0 +1,137 @@
+        file_watcher = FileWatcher(paths=list(model_paths))
+        _logger.info("Monitoring model paths for changes.")
+        async for event in file_watcher:
+            if event.type in (EventType.CREATE, EventType.MODIFY):
```
You can pass the events to watch for to the `FileWatcher()` constructor, then you don't need to check in here.

llucax commented on src/frequenz/sdk/ml/_model_manager.py:
```diff
@@ -0,0 +1,137 @@
+        """Monitor model file paths and reload models as necessary."""
+        model_paths = [model.path for model in self._models.values()]
+        file_watcher = FileWatcher(paths=list(model_paths))
+        _logger.info("Monitoring model paths for changes.")
```
```suggestion
        _logger.info("%s: Monitoring model paths for changes.", self.name)
```

llucax commented on src/frequenz/sdk/ml/_model_manager.py:
```diff
@@ -0,0 +1,137 @@
+        _logger.info("Monitoring model paths for changes.")
+        async for event in file_watcher:
+            if event.type in (EventType.CREATE, EventType.MODIFY):
+                _logger.info("Model file %s modified, reloading...", event.path)
```
```suggestion
                _logger.info("%s: Model file %s modified, reloading...", self.name, event.path)
```

llucax commented on src/frequenz/sdk/ml/_model_manager.py:
```diff
@@ -0,0 +1,137 @@
+        if not self.is_running:
+            task = asyncio.create_task(self.run())
+            self._tasks.add(task)
+            _logger.info("Started ModelManager service with task %s", task)
```
```suggestion
            _logger.info("%s: Started ModelManager service with task %s", self.name, task)
```

llucax commented on src/frequenz/sdk/ml/_model_manager.py:
```diff
@@ -0,0 +1,137 @@
+            if model.path == path:
+                try:
+                    model.data = self._load(path)
+                    _logger.info("Successfully reloaded model from %s", path)
```
Use `self.name` here and in all other logging too. We might want to add some method to background service to do this automatically somehow, but that's for a separate issue.

llucax commented on src/frequenz/sdk/ml/_model_manager.py:
```diff
@@ -0,0 +1,137 @@
+                try:
+                    model.data = self._load(path)
+                    _logger.info("Successfully reloaded model from %s", path)
+                except Exception as e:  # pylint: disable=broad-except
+                    _logger.error("Failed to reload model from %s: %s", path, e)
```
```suggestion
                except Exception:  # pylint: disable=broad-except
                    _logger.exception("Failed to reload model from %s", path)
```

llucax commented on src/frequenz/sdk/ml/_model_manager.py:
```diff
@@ -0,0 +1,137 @@
+                except Exception as e:  # pylint: disable=broad-except
+                    _logger.error("Failed to reload model from %s: %s", path, e)
+
+    async def stop(self, msg: str | None = None) -> None:
+        """Stop all model monitoring tasks with enhanced exception handling.
+
+        Args:
+            msg: An optional message to log when stopping the service.
+        """
+        _logger.info("Stopping ModelManager service: %s", msg)
+        # Attempt to cancel all running tasks
+        for task in list(self._tasks):
+            task.cancel()
+            try:
+                await task  # Wait for task to be cancelled
+            except CancelledError:
+                _logger.info("Task %s cancelled successfully", task)
+            except Exception as e:  # pylint: disable=broad-except
+                _logger.error("Error while cancelling task %s: %s", task, e)
+
+        # Call the parent stop method if it handles additional teardown
+        try:
+            await super().stop(msg)
+        except Exception as e:  # pylint: disable=broad-except
+            _logger.error("Error during stop in superclass: %s", e)
+
```
This is done automatically by `BackgroundService`.

```suggestion
```

llucax commented on src/frequenz/sdk/ml/_model_manager.py:
```diff
@@ -0,0 +1,130 @@
+        super().__init__()
+        self._models: dict[str, _Model[T]] = {}
+        self.model_paths = model_paths
+        self._service_name = name
```
And remove this.
```suggestion
```

idlir-shkurti-frequenz replied:
Ahh ok I see. I wasn't sure how this worked

idlir-shkurti-frequenz replied:
Done!

llucax replied:
> Ahh ok I see. I wasn't sure how this worked

Did you check the example at https://frequenz-floss.github.io/frequenz-sdk-python/v0.25/reference/frequenz/sdk/actor/#frequenz.sdk.actor.BackgroundService?

llucax replied:
If you did and it was still unclear we might need to improve the docs!

llucax commented on src/frequenz/sdk/ml/_model_manager.py:
```diff
@@ -0,0 +1,130 @@
+class ModelManager(BackgroundService, Generic[T]):
+    """Load, update, monitor and retrieve machine learning models."""
+
+    def __init__(self, model_paths: dict[str, Path], name: str = "ModelManager"):
```
I'd leave the default as `None`, the `BackgroundService` can deal with that ("If `None`, `str(id(self))` will be used." [^1]). Also, `str(x)` will use the class name too, so if you use the class name it will look very redundant, like `ModelManager(ModelManager)`.

Accepting a name is only to allow users to customize this, in case they instantiate more than one model manager, it's more so for example, peak shaving can do `mm = ModelManager(paths, name="PeakShaving")`. I would also make `name` a keyword-only argument, like in `BackgroudService`, to avoid confusion:

```suggestion
    def __init__(self, model_paths: dict[str, Path], *, name: str | None = None):
```

[^1]: https://frequenz-floss.github.io/frequenz-sdk-python/v0.25/reference/frequenz/sdk/actor/#frequenz.sdk.actor.BackgroundService.__init__

idlir-shkurti-frequenz replied:
Yes that makes sense. I will change it!

idlir-shkurti-frequenz replied:
Done!

llucax commented on src/frequenz/sdk/ml/_model_manager.py:
```diff
@@ -0,0 +1,130 @@
+                "%s: Model file %s modified, reloading...",
+                self._service_name,
+                event.path,
+            )
```
This might be misleading, maybe this?
```suggestion
            _logger.info(
                "%s: Reloading model from file %s due to a %s event...",
                self.name,
                event.path,
                event.type.name,
            )
```

idlir-shkurti-frequenz replied:
Done!

llucax commented on src/frequenz/sdk/ml/_model_manager.py:
```diff
@@ -0,0 +1,130 @@
+                "%s: Started ModelManager service with task %s",
+                self._service_name,
+                task,
+            )
```
The name is provided by the `BackgroundService` [`name` property](https://frequenz-floss.github.io/frequenz-sdk-python/v0.25/reference/frequenz/sdk/actor/#frequenz.sdk.actor.BackgroundService.name):
```suggestion
            _logger.info(
                "%s: Started ModelManager service with task %s",
                self.name,
                task,
            )
```

(same change for all other logging sites)

idlir-shkurti-frequenz replied:
Done!

llucax reviewed (commented):
Sorry, a few more comments, mostly about how to subclass `BackgroundService` :grimacing:

llucax reviewed (dismissed):
Thanks for the patience!

llucax commented on pyproject.toml:
```diff
@@ -96,6 +96,7 @@ dev-pytest = [
   # For checking docstring code examples
   "frequenz-sdk[dev-examples]",
   "hypothesis == 6.100.0",
+  "sybil == 6.0.3",
```
This should not be necessary anymore, we fixed it in repo-config.

llucax commented on src/frequenz/sdk/ml/_model_manager.py:
```diff
@@ -0,0 +1,130 @@
+        Raises:
+            FileNotFoundError: If the model file does not exist.
+        """
+        if not path.exists():
+            raise FileNotFoundError(f"The model path {path} does not exist.")
```
I would just let `open()` raise the error:
```suggestion
```

If you want to produce a more descriptive/specific error, I suggest creating a custom exception instead and chaining it to the orignal exception so users can get the full context:

```py
class ModelError(Exception):
    # You might want to actually add more details here, like `model_key` and `model_path`
    pass
```    
```suggestion
        try:
            with path.open("rb") as file:
                return cast(T, pickle.load(file))
        except Exception as err:
            raise ModelError(f"Could not load model {model_key} from {model_path}: {e}.", model_key, model_path) from err
```

idlir-shkurti-frequenz replied:
Done!

llucax reviewed (commented):
One last comment while we wait for the release of `rc6.post1` :)

cwasicki commented:
@idlir-shkurti-frequenz Could you rebase this PR please?

idlir-shkurti-frequenz commented:
> @idlir-shkurti-frequenz Could you rebase this PR please?

Done! Also addressed the custom exception point suggested by Luca.

shsms commented:
Looks like the commit wasn't signed, someone might have to create a new PR, ideally with a shorter commit message.

cwasicki commented:
Ok, then I will do some testing with this and re-create it afterwards.

cwasicki commented:
https://github.com/frequenz-floss/frequenz-sdk-python/issues/954

llucax commented:
> Looks like the commit wasn't signed, someone might have to create a new PR, ideally with a shorter commit message.

Didn't get what's going on, it needs to be re-created because Idlir can't do it himself now? We can still amend this PR and force-push to it. Also I don't get the comment about the commit message length.

shsms commented:
> Also I don't get the comment about the commit message length.

a shorter one that doesn't fold on gh would be nice.

llucax commented:
Oh, OK, so unrelated to the signing part, I thought it was related. :+1:

cwasicki commented:
Created a backup here: https://github.com/frequenz-floss/frequenz-sdk-python/pull/968
